pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,531 @@
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/labels"
+	lb "github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/set"
+	"github.com/cilium/cilium/pkg/u8proto"
+)
+
+var (
+	log                 = logging.DefaultLogger.WithField(logfields.LogSubsys, "redirectpolicy")
+	localRedirectSvcStr = "-local-redirect"
+)
+
+type eventRequests struct {
+	forService map[string]*LRPConfig
+}
+
+type svcManager interface {
+	DeleteService(frontend lb.L3n4Addr) (bool, error)
+	UpsertService(*lb.SVC) (bool, lb.ID, error)
+}
+
+type endpointManager interface {
+	GetEndpoints() []*endpoint.Endpoint
+}
+
+// RedirectPolicyManager manages configurations related to Cilium Local Redirect Policies
+// that enable redirecting traffic from the specified frontend to a node-local
+// backend pod selected based on the backend configuration. For every redirect
+// policy configuration, it plumbs datapath with a frontend that has at least one
+// node-local backend.
+// It keeps track of add/delete events for resources like CLRP, Endpoint and Service.
+// While processing different events for a policy configuration, a notify method
+// is called that checks if a policy config is ready for insertion in datapath.
+type RedirectPolicyManager struct {
+	// Stores mapping of all the current redirect policy frontends to their
+	// respective policies
+	// Frontends are namespace agnostic.
+	frontends map[string]string
+	// Stores mapping of redirect policy serviceID to the policyID for
+	// easy lookup in policyConfigs
+	services map[string]policyID
+	// Stores mapping of pod to redirect policy IDs that select this pod
+	podPolicy map[string][]podPolicyInfo
+	// Stores redirect policy configs indexed by policyID
+	policyConfigs map[policyID]*LRPConfig
+
+	// Mutex to protect against concurrent access to the maps
+	lock.RWMutex
+
+	// Service handler to manage service entries corresponding to redirect policies
+	svcManager svcManager
+	// Endpoint Manager to keep track of endpoints and their metadata
+	epManager endpointManager
+
+	// Tracks pending requests to handle events that update redirect policy configs
+	pendingEvents eventRequests
+}
+
+func (rpm *RedirectPolicyManager) storePolicyConfig(config LRPConfig) {
+	_, ok := rpm.policyConfigs[config.getPolicyID()]
+	if !ok {
+		rpm.policyConfigs[config.getPolicyID()] = &config
+	}
+
+	if config.frontend != nil {
+		rpm.frontends[config.frontend.Hash()] = config.name
+	}
+
+	if config.serviceID != nil {
+		rpm.services[config.serviceID.String()] = config.getPolicyID()
+	}
+}
+
+func (rpm *RedirectPolicyManager) updatePolicyConfigFrontend(config *LRPConfig, frontend *frontend, isUpdate bool) {
```

Comment:
This function looks a little odd at first glance, why pass `isUpdate` as a bool if there's effectively two different functions here, an `Upsert...()` and a `Delete...()`?